### DISCUSSION TITLE
Register a middleware for specific service in V5

### DISCUSSION THREAD

**Discussion by barbeblanche96** [feathersjs/feathers#3178] [Q&A]
How can I register a middleware for specific service like with v4. 

```
  app.use('/download-service', new DownloadService(options, app), function(req, res) {
      console.log(res.data);
      const file = fs.readFileSync('./public/'+res.data.path, 'utf8');
      res.type(res.data.type);
      res.end(file);
    });

```
I want to reproduce this with koa in V5, ie register a middleware before and after a specific service not globally like explain in documentation.

**Comment by daffl** [comment]
Whoops I forgot to add this to the docs (now [documented here](https://dove.feathersjs.com/api/koa.html#service-middleware). It is possible to pass a `koa.before` option with custom service middleware when registering the service:

```ts
app.use('/todos', new TodoService(), {
  koa: {
    before: [
      async (ctx, next) => {
        ctx.feathers // data that will be merged into sevice `params`

        // This will run all subsequent middleware and the service call
        await next()

        // Then we have additional properties available on the context
        ctx.hook // the hook context from the method call
        ctx.body // the return value
      }
    ]
  }
})
```

**Comment by SYBIOTE** [comment]
can i register an after service middleware as well?

**Comment by daffl** [comment]
A Koa middleware does both, before and after. You can do things before and after `await next()` as in the comment above. Express middleware registration is documented in https://feathersjs.com/api/express.html#custom-service-middleware

**Comment by SYBIOTE** [comment]
I want to do it with a service object 

Is this syntax valid ?

async function redirect(ctx, next){}

app.use(path, new service(), {koa:{after:[redirect]}})

Edit :
from what your saying , i just need to call redirect after the next() call?

**Reply by daffl** [reply]
An easier way for redirects is setting 

```ts
context.http = {
  ...context.http,
  redirect: 'location'
}
```

In a hook. Also see https://feathersjs.com/api/hooks.html#context-http

**Reply by SYBIOTE** [reply]
hey thanks a lot, I looked into the code base and "after" service middle ware were defined so I used those
